pkg/proxy/dns/store.go

```diff
@@ -0,0 +1,241 @@
+package dns
+
+import (
+	"fmt"
+	"hash/fnv"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/golang/glog"
+
+	skymsg "github.com/skynetservices/skydns/msg"
+)
+
+type dnsStore struct {
+	cache.BulkIndexer
+	endpoints  client.EndpointsNamespacer
+	headlessMu sync.Mutex
+	isHeadless map[string]bool
+}
+
+type dnsEntry struct {
+	Namespace string
+	Name      string
+	skymsg.Service
+}
+
+var _ cache.Store = &dnsStore{}
+
+func makeSkyDNSService(host string, port int, key string, group string) skymsg.Service {
+	return skymsg.Service{
+		Host: host,
+		Port: port,
+
+		Priority: 10,
+		Weight:   10,
+		Ttl:      30,
+
+		Text: "",
+		Key:  key,
+
+		Group: group,
+	}
+}
+
+func makeEntriesForService(service *api.Service) []interface{} {
+	entries := make([]interface{}, len(service.Spec.Ports)+1)
+	srvPath := fmt.Sprintf(srvPathFormat, service.Name, service.Namespace)
```

Comment:
It looks like you're materializing a lot of different entries in the cache per service, which I don't know why that is necessary.